[Security Solution] Use AST utils from @kbn/esql-ast for ES|QL rule type query parsing (#9282)

e40pud opened the pull request:
## Summary

Addresses https://github.com/elastic/security-team/issues/9282

With these changes we utilise AST based utils to do ES|QL query validation. This allows us to recognise and display syntax errors. Syntax errors have higher priority than the rest of the validation errors.

Validation errors priorities from top to bottom:
1. Syntax error
2. Missing metadata for non-aggregating queries
3. Missing data source and/or data fields
4. Missing `_id` column requested for non-aggregating queries via metadata operator

These priorities define the sequence in which we display errors to the user. If there are several errors detected, that the one with higher priority will be shown.

https://github.com/user-attachments/assets/cef88c60-b0a4-413e-885a-b619773fd853

### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
  * [Integration tests](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6681) (100 ESS, 100 Serverless)
  * [Cypress DE tests](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6712) (100 ESS, 100 Serverless)

e40pud committed:
[Security Solution] Use AST utils from @kbn/esql-ast for ES|QL rule type query parsing (#9282)

e40pud commented:
/ci

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/logic/esql_validator.ts:
```diff
@@ -9,6 +9,8 @@ import { isEmpty } from 'lodash';
+import { getAstAndSyntaxErrors } from '@kbn/esql-ast';
+import { validateQuery } from '@kbn/esql-validation-autocomplete';
 import { KibanaServices } from '../../../common/lib/kibana';
 
```
there are multiple places when we need to use AST parser, to address https://github.com/elastic/security-team/issues/9282:

https://github.com/elastic/kibana/blob/8.15/x-pack/plugins/security_solution/public/detection_engine/rule_creation/logic/esql_validator.ts#L41

https://github.com/elastic/kibana/blob/8.15/packages/kbn-securitysolution-utils/src/esql/compute_if_esql_query_aggregating.ts

What is the intended scope of this PR?

e40pud replied:
Good points!

Regarding the scope: we need to be able to a) get syntax errors if any while validating the query; b) switch to AST based utils to do query parsing.

Addressed points mentioned above in the new commit.

e40pud replied:
done

e40pud committed:
Review feedback

e40pud commented:
/ci

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

e40pud commented:
/ci

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/logic/esql_validator.ts:
```diff
@@ -34,11 +38,40 @@ const constructValidationError = (error: Error) => {
-export const computeHasMetadataOperator = (esqlQuery: string) => {
-  return /(?<!\|[\s\S.]*)\s*metadata[\s\S.]*_id[\s\S.]*/i.test(esqlQuery?.split('|')?.[0]);
+export const computeHasMetadataOperator = (ast: ESQLAst) => {
+  const commands = ast.filter((a) => a.type === 'command');
+  const fromCommand = commands.find((command) => command.name === 'from');
```
can we juse use one `find` here?
```suggestion
  const fromCommand = ast.find((command) => a.type === 'command' && a.name === 'from');
```

vitaliidm commented on packages/kbn-securitysolution-utils/src/esql/compute_if_esql_query_aggregating.ts:
```diff
@@ -6,11 +6,19 @@
+import { ESQLAst, getAstAndSyntaxErrors } from '@kbn/esql-ast';
+
+export const isAggregatingQuery = (ast: ESQLAst): boolean => {
+  const commands = ast.filter((a) => a.type === 'command');
+  return !!commands.find((command) => command.name === 'stats');
```
same as [here](https://github.com/elastic/kibana/pull/189780/files#r1701932661)

e40pud replied:
done

e40pud committed:
Merge branch 'main' into security/feature/esql-ast-validator

e40pud committed:
Fix tests

e40pud commented:
/ci

e40pud committed:
Add new tests & Review feedback

e40pud commented:
/ci

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-engine (Team:Detection Engine)

vitaliidm commented on x-pack/plugins/security_solution/public/detection_engine/rule_creation/logic/esql_validator.ts:
```diff
@@ -34,11 +38,39 @@ const constructValidationError = (error: Error) => {
+  }
+
+  // Check whether the `metadata` operator has `_id` argument
+  const idColumnItem = (metadataOption as ESQLCommandOption).args.find(
```
why casting is needed here? 
Given that one of the types there does not have `args` property, it could lead to error of accessing property. Something we had already in the past.

e40pud replied:
Seems like Array.find cannot infer the type in this case. I re-arranged the code.

kibanamachine commented:
## Flaky Test Runner Stats
### 🎉 All tests passed! - [kibana-flaky-test-suite-runner#6681](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6681)
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/configs/ess.config.ts: 100/100 tests passed.
[✅] x-pack/test/security_solution_api_integration/test_suites/detections_response/detection_engine/rule_execution_logic/trial_license_complete_tier/configs/serverless.config.ts: 100/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F189780%2Fhead)

e40pud committed:
Remove casting

e40pud committed:
Merge branch 'main' into security/feature/esql-ast-validator

kibanamachine commented:
## Flaky Test Runner Stats
### 🟠 Some tests failed. - [kibana-flaky-test-suite-runner#6682](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/6682)
[❌] Security Solution Detection Engine - Cypress: 52/100 tests passed.
[❌] [Serverless] Security Solution Detection Engine - Cypress: 81/100 tests passed.

[see run history](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds?branch=refs%2Fpull%2F189780%2Fhead)

vitaliidm reviewed (approved):
LGTM
thanks for you work

e40pud committed:
Fix cypress tests

These changes https://github.com/elastic/kibana/pull/188898 caused new tests to fail

e40pud committed:
Merge branch 'main' into security/feature/esql-ast-validator

e40pud committed:
Merge branch 'main' into security/feature/esql-ast-validator

# Conflicts:
#	x-pack/plugins/security_solution/tsconfig.json

e40pud committed:
Merge branch 'main' into security/feature/esql-ast-validator
